### ISSUE TITLE
Extends Issue >=3.2.8

### ISSUE THREAD

**Issue by danhere** [sass/sass#788]
We're having an issue since version >= 3.2.8 of Sass in which extends are not being compiled correctly.

In the Gumby Framework, we're using the following function:

``` scss
.row {
  @for $i from 1 through $cols{
    @if $i == 1 {
      .one.column {
        width: columns($i);
      }
      .one.columns {
        @extend .one.column
      }
    }
    @else{
      .#{number-as-word($i)}.columns {
        width: columns($i);
      }
    }
  }
}
```

renders out as:

``` css
.sixteen.colgrid .row .one.column{
  width: 4.25532%;
}
```

when it should be:

``` css
.sixteen.colgrid .row .one.column, .sixteen.colgrid .row .one.columns {
  width: 4.25532%;
}
```

This isn't the only place the @extends are getting ignored.

Any ideas?

**Comment by chriseppstein** [comment]
I cannot reproduce this using the latest master code nor Sass 3.2.9.

**Comment by danhere** [comment]
It seems that we can't extend multiple classes in a single extend.
Running `@extend .one.column;` doesn't work but `@extend .one;@extend .column;` does.

`@extend .one.column` works up to 3.2.7.
After that though it no longer works. Did something change? Can you no longer extend chained selectors?

**Comment by cimmanon** [comment]
The code you've provided works fine for me (3.3.0.alpha.175), except for the part where the `columns()` function isn't defined.

``` css
.row .one.column, .row .one.columns {
  width: columns(1);
}
```

**Comment by chriseppstein** [comment]
There is no 3.5.7 release. Assuming you meant 3.2.7.

Here's the test case I'm using:

``` scss
.one { foo: bar; } // shouldn't be extended.
.column { foo: bar; } // shouldn't be extended.
.row {
  .one.column { // should be extended and produce '.row .one.columns'
    width: 250px;
  }
  .one.columns {
    @extend .one.column
  }
}
```

and it produces (with 3.2.9):

``` css
.one { foo: bar; }

.column { foo: bar; }

.row .one.column, .row .one.columns { width: 250px; }
```

Which is what I would expect.

**Comment by danhere** [comment]
Okay so, I boiled it down and this is what I came up with:

```
.one.column{
    foo: bar;
}

.sixteen.colgrid .row{
    .one.column {
        width: columns(1, true);
      }
      .one.columns {
        @extend .one.column;
      }
}
```

Returns:

```
.one.column, .sixteen.colgrid .row .one.columns {
  foo: bar; }

.sixteen.colgrid .row .one.column, .sixteen.colgrid .row .one.columns {
  width: columns(1, true); }
```

I guess this is correct and we were just relying on an old bug with the @extend function to work? Seems incorrect to me.

**Comment by craigkeller** [comment]
Hey @chriseppstein,

Here is a better example of our issue.

We have this Sass:

```
.mycoolclass {
 .foo.bar {
   width: 99px;
 }
 .foo.bars {
   @extend .foo.bar
 }
}

.morespecificclass .mycoolclass {
 .foo.bar {
   width: 2000px;
 }
 .foo.bars {
   @extend .foo.bar
 }
}
```

We expect that should return:

```
.mycoolclass .foo.bar, .mycoolclass .foo.bars { width: 99px; }

.morespecificclass .mycoolclass .foo.bar, .morespecificclass .mycoolclass .foo.bars { width: 2000px; }
```

But instead, its returning this, which is inconsistent with Sass versions prior to v3.2.8:

```
.mycoolclass .foo.bar, .mycoolclass .foo.bars, .morespecificclass .mycoolclass .foo.bars { width: 99px; }

.morespecificclass .mycoolclass .foo.bar { width: 2000px; }
```

Thoughts?

**Comment by chriseppstein** [comment]
Extend is not scoped. It affects all selectors that reference the simple selector that you are extending.

**Comment by craigkeller** [comment]
Thanks.

Just to confirm – this behavior changed starting in 3.2.8? Meaning that it was scoped in prior versions unintentionally.

I just want to make sure that this will continue to be the case moving forward so that I can make the appropriate adjustments to some conventions that we use in Gumby.

**Comment by chriseppstein** [comment]
Ok. Thanks @craigkeller I see the issue now. Reopening -- I agree this is a regression.

This was changed in 3700884c8c150e61da8d24168d25db34fccbd19b which was first released in 3.2.8.

For what it's worth, as a work around, this code works and I think it more succinctly expresses your particular use case:

``` scss
 .foo.bars {
   @extend .foo.bar
 }

.mycoolclass {
 .foo.bar {
   width: 99px;
 }
}

.morespecificclass .mycoolclass {
 .foo.bar {
   width: 2000px;
 }
}
```

**Comment by craigkeller** [comment]
thanks, @chriseppstein

**Comment by chriseppstein** [comment]
I think I follow what @nex3 was trying to do in that commit but I don't think it quite accomplishes it.

I found some other disconcerting issues when I started poking around.

``` scss
%foo %bar {
  a: b;
}
%foo %bar%baz {
  c: d;
}
.context %foo %bar {
  e: f;
}
%foo .another-context %baz {
  g: h;
}

.parent1 {
  @extend %foo;

  .child1 {
    @extend %bar;
  }
}

.parent2 {
  @extend %foo;

  .child2 {
    @extend %baz;
  }
}
```

Which I expected to generate:

``` css
.parent1 .child1 {
  a: b; }

.parent1 .child1.child2, .parent2 .child1.child2 {
  c: d; }

.context .parent1 .child {
  e: f; }

.parent2 .another-context .child2 {
  g: h;
}
```

But instead I get:

``` css
.parent1 .child1 {
  a: b; }

.context .parent1 .child1, .parent1 .context .parent2 .child1 {
  e: f; }

.parent1 .another-context .parent2 .child2, .parent2 .another-context .child2 {
  g: h; }
```

So we're completely missing the case where the children could be the same element and we've got some extraneous parent selectors getting mixed into the selectors which should be optimized out when the selector gets more complex.

**Comment by nex3** [comment]
@chriseppstein I think only the `c: d` selector in your example is incorrect. For `e: f`, putting `.parent1` before `.context` fulfills `.parent1 .child1 { @extend %bar }` and `.parent2 { @extend %foo }`. Similarly for `g: h`.

I'm working on fixing the incorrect compilation here, though.

**Comment by nex3** [comment]
Just reverting 3700884c8c150e61da8d24168d25db34fccbd19b didn't fix the issue; it looks like 5bc676942daa24e81f4a473894652a18b8b85a41 is also contributing somehow. Reverting both commits simultaneously does fix @craigkeller's issue.

Even going back before 3700884c8c150e61da8d24168d25db34fccbd19b doesn't fix @chriseppstein's `c: d` issue, which suggests that it may have a different root cause.

**Comment by nex3** [comment]
I'm pushing the issue-788 branch with some unit tests because I'm not sure I'm going to be able to solve this immediately and I'm going to be offline for a while very shortly.
